cpp/src/arrow/array/array_binary_test.cc

```diff
@@ -301,14 +301,18 @@ class TestStringArray : public ::testing::Test {
     auto st2 = ValidateFull(1, {0, 4}, "\xf4\x90\x80\x80");
     // Single UTF8 character straddles two entries
     auto st3 = ValidateFull(2, {0, 1, 2}, "\xc3\xa9");
+    // Null characters in the string
+    auto st4 = ValidateFull(1, {0, 4}, "\0\0\0\0");
```

Comment:
Can you explain why this is invalid? Unicode character 0 is a valid unicode character.

Reply:
Also, it works using PyArrow:
```python
>>> arr = pa.array(["\0"])
>>> arr
<pyarrow.lib.StringArray object at 0x7f6befb354b0>
[
  ""
]
>>> arr.to_pylist()
['\x00']
>>> arr.validate(full=True)
>>> 
```

Reply:
You are right, this case is valid. The UTF-8 standard (https://tools.ietf.org/html/rfc3629) states that characters consisting of 1-byte range from 0-7F (Section 3). But a multibyte UTF-8 character will not contain NUL as a continuation byte, the encoding does not allows it.
Valid: '\x00'
Invalid: '\xe10080'

Reply:
Also, note that UTF8 validation is tested more thoroughly in https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/utf8_util_test.cc